crates/configuration/src/version4/mod.rs

```diff
@@ -188,6 +200,85 @@ pub async fn introspect(
     })
 }
 
+/// given scalar type names already in use, introspected tables, and optionally any existing table configuration:
+/// get collections with names guaranteed unique, preserving customized collection and field names if any
+fn get_aliased_tables(
+    type_names: HashSet<TypeName>,
+    tables: metadata::TablesInfo,
+    old_tables: &metadata::TablesInfo,
+) -> metadata::TablesInfo {
+    let mut type_names = type_names;
+    let mut mapped_tables = BTreeMap::new();
+
+    for (collection_name, table_info) in tables.0 {
+        let old_config = old_tables.0.iter().find(|(_, old_table_info)| {
+            old_table_info.table_name == table_info.table_name
+                && old_table_info.schema_name == table_info.schema_name
+        });
+
+        // use the old collection alias if one exists
+        let collection_name = old_config
+            .map_or(&collection_name, |(collection_name, _)| collection_name)
+            .to_owned();
+
+        // add a suffix to the collection name if needed
+        let collection_name = get_unique_collection_name(collection_name, &type_names);
+
+        type_names.insert(collection_name.clone().into_inner().into());
+
+        // if a column has a customized field name, keep it
+        let table_info = metadata::TableInfo {
+            columns: table_info
+                .columns
+                .into_iter()
+                .map(|(field_name, column_info)| {
+                    let field_name = old_config
+                        .and_then(|(_, table_info)| {
+                            table_info
+                                .columns
+                                .iter()
+                                .find(|(_, old_column_info)| {
+                                    old_column_info.name == column_info.name
+                                })
+                                .map(|(field_name, _)| field_name.to_owned())
+                        })
+                        .unwrap_or(field_name);
+
+                    (field_name, column_info)
+                })
+                .collect(),
+            ..table_info
+        };
+
+        mapped_tables.insert(collection_name, table_info);
+    }
+
+    metadata::TablesInfo(mapped_tables)
+}
+
+/// given a collection name and a list of already used type names, get a unique name by adding a suffix if needed
+fn get_unique_collection_name(
+    collection_name: CollectionName,
+    type_names: &HashSet<TypeName>,
+) -> CollectionName {
+    let mut collection_name = collection_name;
+
+    if type_names.contains(&collection_name.clone().into_inner()) {
```

Comment:
The newtype macro implements `AsRef` for borrowing the inner value so you can do,

```suggestion
    if type_names.contains(collection_name.as_ref()) {
```